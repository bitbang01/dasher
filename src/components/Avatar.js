import { oneOfType, oneOf, string, number } from 'prop-types'
import styled from 'react-emotion'

import { radii } from '../theme'

const Avatar = styled.img(props => ({
  width: props.size,
  height: props.size,
  borderRadius: props.shape === 'square' ? radii[0] : '50%',
  verticalAlign: 'middle',
}))

Avatar.propTypes = {
  src: string.isRequired,
  size: oneOfType([string, number]),
  shape: oneOf(['circle', 'square']),
}

Avatar.defaultProps = {
  size: 24,
  shape: 'square',
}

export default Avatar
