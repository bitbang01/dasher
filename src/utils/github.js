import qs from 'qs'

const GITHUB_CLIENT_ID = '799bd70f09884c025750'

export function getGithubAuthUrl({ origin }) {
  const redirectUri = origin + '/callback'

  const queryString = qs.stringify({
    client_id: GITHUB_CLIENT_ID,
    scope: 'repo',
    redirect_uri: redirectUri,
  })

  return `https://github.com/login/oauth/authorize?${queryString}`
}
