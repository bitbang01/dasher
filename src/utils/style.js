import color from 'color-string'
import memoize from 'fast-memoize'

const isNumber = value => typeof value === 'number' && !isNaN(value)

const toUnit = unit => value =>
  isNumber(value) && value !== 0 ? value + unit : value

export const toPx = toUnit('px')

export const toEm = toUnit('em')

export const toMediaQuery = breakpoint =>
  `@media screen and (min-width: ${toEm(breakpoint)})`

export const joinSpacing = (...args) => {
  if (args.length > 4) {
    throw new Error('Too many arguments')
  }

  return args.map(toPx).join(' ')
}

/**
 * Returns the transparent version of the `foreground` color,
 * as if it was overlaid on the `background` color.
 *
 * Inspiration: https://git.io/vpYAv
 *
 * @param {string} foreground
 * @param {string} [background=#fff]
 * @returns {string}
 */
export const toAlpha = memoize(
  (foreground, background = '#fff') => {
    const r = 0
    const g = 1
    const b = 2

    // Convert color string to array (i.e. [r, g, b, a])
    const fgColor = color.get.rgb(foreground)
    const bgColor = color.get.rgb(background)

    // Calculate alpha value
    let alpha = [r, g, b]
      .map(
        channel =>
          (fgColor[channel] - bgColor[channel]) /
          ((0 < fgColor[channel] - bgColor[channel] ? 255 : 0) -
            bgColor[channel]),
      )
      .sort((a, b) => b - a)[0]

    // Keep alpha value between 0 and 1
    alpha = Math.max(Math.min(alpha, 1), 0)

    // Calculate the resulting color
    function processChannel(channel) {
      return 0 === alpha
        ? bgColor[channel]
        : bgColor[channel] + (fgColor[channel] - bgColor[channel]) / alpha
    }

    return color.to.rgb(
      processChannel(r),
      processChannel(g),
      processChannel(b),
      Math.round(alpha * 100) / 100,
    )
  },
  {
    strategy: memoize.strategies.variadic,
  },
)

/**
 * Makes it easier to create glamorous components that
 * accept props to enable/disable styles.
 *
 * Accepts an object where the key is a prop and the value is
 * the styles object that should be applied if that prop is passed.
 * Returns a function which you pass to a glamorousComponentFactory.
 *
 * Source: https://codesandbox.io/s/AGRRMl63 by @kentcdodds
 *
 * @param {Object} styles - An object mapping props to styles
 * @return {Function} - A dynamic styles function
 */
export const propStyles = styles => (props = {}) =>
  Object.keys(props).map(key => {
    if (props[key]) {
      return typeof styles[key] === 'function'
        ? styles[key](props)
        : styles[key]
    }
    return null
  })
