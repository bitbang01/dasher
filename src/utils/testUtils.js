// Named this file `testUtils.js` instead of `test.js` because Jest treats
// `test.js`as a test file
// Source: https://jestjs.io/docs/en/configuration#testmatch-array-string

import { ThemeProvider } from 'emotion-theming'
import React from 'react'
import { render } from 'react-testing-library'
import theme from '../theme'

export const renderWithTheme = children =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>)
