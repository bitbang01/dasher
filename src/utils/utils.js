export const get = (path, object) =>
  path.reduce(
    (acc, property) => (acc && acc[property] ? acc[property] : null),
    object,
  )

export const reorder = (list, from, to) => {
  const listCopy = [...list]
  const [removed] = listCopy.splice(from, 1)
  listCopy.splice(to, 0, removed)
  return listCopy
}
