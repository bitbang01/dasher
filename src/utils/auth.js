import { gql } from 'apollo-boost'

import { client } from '../components/Apollo'

export const AUTH_TOKEN_KEY = 'auth_token'
export const GITHUB_TOKEN_KEY = 'github_token'

const AUTHENTICATE_MUTATION = gql`
  mutation AuthenticateMutation($githubCode: String!) {
    authenticate(githubCode: $githubCode) {
      token
      githubToken
    }
  }
`

export async function logIn(githubCode) {
  const { data } = await client.mutate({
    mutation: AUTHENTICATE_MUTATION,
    variables: { githubCode },
  })

  setAuthToken(data.authenticate.token)
  setGithubToken(data.authenticate.githubToken)
}

export function logOut() {
  removeAuthToken()
  removeGithubToken()
}

export function isLoggedIn() {
  return Boolean(getAuthToken())
}

export function setAuthToken(token) {
  localStorage.setItem(AUTH_TOKEN_KEY, token)
}

export function getAuthToken() {
  return localStorage.getItem(AUTH_TOKEN_KEY)
}

export function removeAuthToken() {
  localStorage.removeItem(AUTH_TOKEN_KEY)
}

export function setGithubToken(githubToken) {
  localStorage.setItem(GITHUB_TOKEN_KEY, githubToken)
}

export function getGithubToken() {
  return localStorage.getItem(GITHUB_TOKEN_KEY)
}

export function removeGithubToken() {
  localStorage.removeItem(GITHUB_TOKEN_KEY)
}
