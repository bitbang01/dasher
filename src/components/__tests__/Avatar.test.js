import React from 'react'
import { renderWithTheme } from '../../utils/testUtils'
import Avatar from '../Avatar'

it('renders without crashing', () => {
  const { container } = renderWithTheme(
    <Avatar src="http://via.placeholder.com/100x100" alt="placeholder" />,
  )
  expect(container.firstChild).toMatchSnapshot()
})

it('renders with custom size (number)', () => {
  const { container } = renderWithTheme(
    <Avatar
      src="http://via.placeholder.com/100x100"
      alt="placeholder"
      size={32}
    />,
  )
  expect(container.firstChild).toMatchSnapshot()
})

it('renders with custom size (string)', () => {
  const { container } = renderWithTheme(
    <Avatar
      src="http://via.placeholder.com/100x100"
      alt="placeholder"
      size="4rem"
    />,
  )
  expect(container.firstChild).toMatchSnapshot()
})
