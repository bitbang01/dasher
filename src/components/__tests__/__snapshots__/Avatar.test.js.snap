// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with custom size (number) 1`] = `
.emotion-0 {
  height: 32px;
  width: 32px;
  border-radius: 4px;
  vertical-align: middle;
}

<img
  src="http://via.placeholder.com/100x100"
  alt="placeholder"
  class="emotion-0"
/>
`;

exports[`renders with custom size (string) 1`] = `
.emotion-0 {
  height: 4rem;
  width: 4rem;
  border-radius: 4px;
  vertical-align: middle;
}

<img
  src="http://via.placeholder.com/100x100"
  alt="placeholder"
  class="emotion-0"
/>
`;

exports[`renders without crashing 1`] = `
.emotion-0 {
  height: 24px;
  width: 24px;
  border-radius: 4px;
  vertical-align: middle;
}

<img
  src="http://via.placeholder.com/100x100"
  alt="placeholder"
  class="emotion-0"
/>
`;
